onnx_chainer/graph.py

```diff
@@ -0,0 +1,179 @@
+import collections
+from collections import OrderedDict
+import heapq
+
+import chainer
+
+from onnx_chainer.functions.converter import FunctionConverterParams
+from onnx_chainer import onnx_helper
+
+
+class Graph(object):
+
+    def __init__(
+            self, context, converters, opset_version, network_outputs):
+        self.context = context
+        self.converters = converters
+
+        self.graph = []
+        # Converter nodes keyed by "number:func_name"
+        self.converted_nodes = OrderedDict()
+        self.func_name_counts = collections.defaultdict(int)
+        self.inputs = {}  # Input `Variable` objects keyed by string IDs
+        self.additional_parameters = []
+        self.specified_opset_version = opset_version
+        self.network_outputs = network_outputs
+
+        self.function_nodes = self._build_computational_graph(
+            network_outputs.values())
+
+    def _build_computational_graph(self, outputs):
+        cands = []
+        seen_edges = set()
+        function_nodes = OrderedDict()
+        push_count = [0]
+
+        def add_cand(cand):
+            heapq.heappush(cands, (-cand.rank, push_count[0], cand))
+            push_count[0] += 1
+
+        for o in outputs:
+            if isinstance(o, chainer.Variable):
+                o = o.node
+            add_cand(o)
+
+        while cands:
+            _, _, cand = heapq.heappop(cands)
+            if not isinstance(cand, chainer.variable.VariableNode):
+                raise NotImplementedError(
+                    'ONNX-Chainer does not support node type {}'.format(
+                        type(cand)))
+            creator = cand.creator_node
+            if creator is None:
+                continue
+            if (creator, cand) in seen_edges:
+                continue
+            assert isinstance(creator, chainer.FunctionNode)
+            creator_id = id(creator)
+            if creator_id in function_nodes:
+                continue
+            seen_edges.add((creator, cand))
+            function_nodes[creator_id] = creator
+
+            for input_ in creator.inputs:
+                if input_ is creator:
```

Comment:
Wow. Does this really happen? Could you add a comment for what kind of case this check is necessary?